<a href="https://colab.research.google.com/github/Marlon-Sbardelatti/machine-learning/blob/feature%2Ftrabalho-04/trabalho-04/neural_network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

> **Equipe**: 01<br>
> **Estudantes**: Cristina Siewert Jansen, Marlon Sbardelatti e Sofia Sousa Lindner

# **REDES NEURAIS ARTIFICIAIS**

# PARTE 1: Classificação Binária - Diagnóstico de Câncer de Mama

Base de dados: Breast Cancer Wisconsin (Diagnostic) -
https://archive.ics.uci.edu/dataset/17/breast+cancer+wisconsin+diagnostic

In [2]:
# Instala bilbiotecas Keras e Tensorflow
!pip install scikeras==0.13.0

In [3]:
!pip install tensorflow==2.21.0 scikit-learn==1.9.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 572.6/572.6 MB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 106.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 108.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 327.1/327.1 kB 26.7 MB/s eta 0:00:00
  Attempting uninstall: protobuf
    Found existing installation: protobuf 5.29.6
    Uninstalling protobuf-5.29.6:
      Successfully uninstalled protobuf-5.29.6
  Attempting uninstall: h5py
    Found existing installation: h5py 3.16.0
    Uninstalling h5py-3.16.0:
      Successfully uninstalled h5py-3.16.0
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.6.1
    Uninstalling scikit-learn-1.6.1:
      Successfully uninstalled scikit-learn-1.6.1
  Attempting uninstall: tensorflow
    Found existing installation: tensorflow 2.20.0
    Uninstalling tensorflow-2.20.0:
      Successfully uninstalled tensorflow-2.20.0
ERROR: pip's dependency re

In [1]:
# após instalar as bibliotecas, é necessário reiniciar o notebook
# Para isso, rode esta célula sozinha antes de prosseguir.
exit()

In [1]:
import pandas as pd
import tensorflow as tf
import sklearn
import scikeras

In [2]:
# Exibe versões das bibliotecas instaladas
pd.__version__, tf.__version__, sklearn.__version__, scikeras.__version__

('2.2.2', '2.21.0', '1.9.0', '0.13.0')

### Carregar os dados

**1) Carregue a base de dados, faça a divisição de treino e teste (para isso, utilize a função train_test_split do sklearn), como o tamanho da base de teste de 0.25.**

In [3]:
from sklearn.model_selection import train_test_split

# Lê arquivos CSV
breast_data = pd.read_csv('dados_breast.csv')
breast_labels = pd.read_csv('rotulos_breast.csv')

# Converte para ndarray em X e y
X = breast_data.values
y = breast_labels.values

# Separa dataset em conjunto de treino e teste (25%)
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.25,
)

# Exibe dimensões do conjunto de treino
X_train.shape, y_train.shape

((426, 30), (426, 1))

In [4]:
# Exibe dimensões do conjunto de teste
X_test.shape, y_test.shape

((143, 30), (143, 1))

Ambos os conjuntos têm duas dimensões: número de amostras, número de colunas. No caso do eixo X, as 30 colunas representam features distintas das amostras de entrada. No caso do eixo y, há apenas uma coluna, que é associada ao rótulo de cada amostra.

A separação de 426 amostras para treino e 143 amostras para teste confirma a separação escolhida de 25%, visto que 426 + 143 = 569 (total) e 569 × 0.25 ≈ 143.

### Estrutura da Rede Neural Artificial e Teste

In [5]:
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import cross_val_score
from tensorflow.keras.models import Sequential
import tensorflow as tf
from tensorflow.keras import backend as k

**2) Crie a RNA com as seguintes configurações:**

**a) uma camada de entrada de com 30 neurônios;**

**b) uma camada oculta densa com 16 neurônios. Porque a camada oculta tem essa quantidade de neurônios?**

**c) adicione na camada oculta a função de ativação relu e inicialize os pesos de utilize o Random uniform initializer (https://www.tensorflow.org/api_docs/python/tf/keras/initializers);**

**d) adicione a camada de saíde com a função de ativação sigmoid. Porque foi utilizada esta função para a saída da RNA?**

In [6]:
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.initializers import RandomUniform

ENTRY_LAYER_SIZE = 30
HIDDEN_LAYER_SIZE = 16

# d) Inicializador dos pesos aleatórios da camada oculta
initializer = RandomUniform(minval=0.0, maxval=1.0)

neural_network = Sequential([
    # a) Camada de Entrada
    Input(shape=(ENTRY_LAYER_SIZE,)),

    # b) Camada Oculta
    Dense(units=HIDDEN_LAYER_SIZE, activation='relu', kernel_initializer=initializer),

    # d) Camada de Saída: Sigmoid para classificação binária
    Dense(units=1, activation='sigmoid')
])

classifier = KerasClassifier(
    model=neural_network,
    verbose=0
)

**b) uma camada oculta densa com 16 neurônios. Porque a camada oculta tem essa quantidade de neurônios?**<br>
Porque o número de neurônios da camada oculta é: nº de neurônios de entrada + nº de saídas / 2.<br>
Dessa forma, temos: (30 + 2) / 2 = 32 / 2 = 16.


**d) adicione a camada de saída com a função de ativação sigmoid. Porque foi utilizada esta função para a saída da RNA?**<br>
Porque a classificação é binária, ou seja, só existem duas classes possíveis: 0 (Sem câncer) ou 1 (Com câncer). Sendo assim, a função sigmoide por padrão utiliza o limiar 0.5, o que permite a classificação em exatas duas classes.

**3) Porque utilizamos a classe chamada Sequential para a RNA?**<br>
A classe Sequential permite a definição de camadas sequenciais à rede, o que altera completamente sua arquitetura e comportamento, visto que o resultado de cada camada é utilizado na próxima. Ou seja, para definirmos que os resultados perpassam camada de entrada, oculta e saída sequencialmente, precisamos utilizar o Sequential.

**4) A partir da RNA gerada, explique o que são os valores apresentados na tabela da rede_neural.summary()**<br>
São apresentadas as camadas oculta e de saída da rede, bem como a quantidade de parâmetros e saídas de cada uma. Naturalmente, a camada oculta produz 16 saídas porque contém 16 neurônios, enquanto a camada de saída produz apenas uma: a classe.<br>
Já a coluna de parâmetros exibe quantos parâmetros treináveis existem na camada. Dado que o cálculo dos parâmetros é (nº de entradas x nº de neurônios) + biases, a camada oculta tem 496 pois (31 x 16) + 0 = 496, enquanto a de saída tem 17 pois (16 x 1) + 1 = 17.

In [8]:
neural_network.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 513 (2.00 KB)

 Trainable params: 513 (2.00 KB)

 Non-trainable params: 0 (0.00 B)

**5) Adicione um otimizador Adam (https://www.tensorflow.org/api_docs/python/tf/keras/optimizers) e especifique a classe loss binário - binary crossentropy (https://www.tensorflow.org/api_docs/python/tf/keras/losses) e a classe metrics para utilizar a métrica de avaliação de acurácia binária (https://www.tensorflow.org/api_docs/python/tf/keras/metrics)**

In [10]:
# Compila rede com função de perda binária e acurácia binária como métrica de validação
neural_network.compile(optimizer='adam', loss='binary_crossentropy', metrics=['binary_accuracy'])

**6) Para que servem os otimizadores? Como o otimizador Adam funciona?**

Os otimizadores têm a função de atualizar os pesos da rede neural ao longo do treinamento, com o objetivo de minimizar a função de perda e melhorar a precisão das previsões.

O algoritmo Adam (Adaptive Moment Estimation) ajusta automaticamente as taxas de aprendizado para cada parâmetro da rede neural combinando duas técnicas clássicas: Momentum e RMSprop. Ele utiliza estimativas da média dos gradientes e da variância dos gradientes para adaptar automaticamente a taxa de aprendizado para cada peso individual do modelo. Essa abordagem proporciona uma convergência mais rápida e estável, tornando o Adam um dos métodos de otimização mais utilizados em problemas de aprendizado profundo.

**7) Depois de estruturados os parâmtros da RNA, utilize a função .fit para fazer o treinamento da rede. Como foi utilizado o otimizador Adam, e ele é baseado na descida do gradiente estocástica, é possível definir a quantidade de registros que serão enviados para a RNA, isto é, em cada batch serão utilizados 10 registros. Quantos batches serão utilizados ao total?**

Como o número de registros do conjunto de treinamento é 426 e o `batch_size` é 10, o total de batches **por época** pode ser dado por 426 / 10 ≈ 43. Finalmente, considerando que o treinamento definiu o valor arbitrário de 100 épocas, serão utilizados 43 × 100 = 4300 batches ao longo de todo o treinamento.

**8) Por fim, defina o número de épocas em que ocorre o treinamento igual a 100.**

In [11]:
# Realiza treinamento do modelo com 100 épocas
neural_network.fit(X_train, y_train, batch_size=10, epochs=100)

Epoch 1/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 2s 21ms/step - binary_accuracy: 0.6502 - loss: 132.8360
Epoch 2/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6221 - loss: 74.9856
Epoch 3/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6009 - loss: 61.2179
Epoch 4/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6385 - loss: 38.0046
Epoch 5/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6033 - loss: 26.5005
Epoch 6/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5915 - loss: 22.8324
Epoch 7/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5915 - loss: 19.2905
Epoch 8/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6268 - loss: 14.4833
Epoch 9/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6244 - loss: 13.0635
Epoch 10/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6573 - loss: 10.9621
Epoch 11/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accura

**9) Crie uma variável chamada previsoes para realizar a previsão dos dados de teste (X_teste) O resultado da rede deve ser um valor entre 0 e 1. Porque isso acontece?** <br>

Como estamos usando a função de ativação sigmoid na camada de saída da rede, o resultado da previsão será uma probabilidade entre 0 (0%) e 1 (100%). Ainda que o problema seja de classificação binária, o threshold de 0.5 só é utilizado após previsão, de modo que saídas > 0.5 = 1 e saídas <= 0.5 = 0. Isto porque, se não fosse um problema do gênero, a utilização de outro threshold permitira dividir em mais classes o resultado.

In [12]:
# Realiza previsão para conjunto de teste
predictions = neural_network.predict(X_test)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 54ms/step


**10) Converta o resultado que está em probabilidade para valores binários com um limiar (treshold) de 0.5**.

In [15]:
# Conversão das previsões para classes binárias: 0 (False) ou 1 (True)
predictions = predictions > 0.5

In [16]:
predictions

array([[False],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [False],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [ True],
       [ True],
       [

**9) Explique o resultado da RNA.** <br>

A RNA, apesar de ter utilizado 100 épocas para treinamento, apresentou um resultado intermediário de acurácia, de 86,71%, e perda, de 2. Provavelmente isso se deve à arquitetura da rede, que utiliza apenas uma camada oculta (Perceptron). A taxa relativamente baixa de perda revela que o otimizador foi importante durante o processo de treinamento, ainda que talvez não tenha a configuração ideal para alcançar melhor eficiência da rede.

In [17]:
neural_network.evaluate(X_test, y_test)

5/5 ━━━━━━━━━━━━━━━━━━━━ 1s 71ms/step - binary_accuracy: 0.8671 - loss: 2.0709


[2.070936918258667, 0.867132842540741]

## Camadas e Otimização da RNA

**10) Adicione mais uma camada para oculta densa com 16 neurônio com a função de ativação relu e inicialize os pesos de utilize o Random uniform initializer. Qual o total de parâmetros da RNA agora?**<br>

O total de parâmetros agora é de 785

In [ ]:
neural_network = Sequential([
    Input(shape=(ENTRY_LAYER_SIZE,)),

    Dense(units=HIDDEN_LAYER_SIZE, activation='relu', kernel_initializer=initializer),
    Dense(units=HIDDEN_LAYER_SIZE, activation='relu', kernel_initializer=initializer),

    Dense(units=1, activation='sigmoid')
])

classifier = KerasClassifier(
    model=neural_network,
    verbose=0
)

neural_network.summary()


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                 │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 16)             │           272 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 785 (3.07 KB)

 Trainable params: 785 (3.07 KB)

 Non-trainable params: 0 (0.00 B)

**11) Abaixo são adicionados os parâmetros do otimizador, que são a taxa de aprendizado e o clipvalue. O que eles fazem?** <br>

A taxa de aprendizado (learning rate) determina o tamanho dos ajustes realizados nos pesos da rede a cada atualização. Valores maiores permitem um aprendizado mais rápido, mas podem fazer com que o modelo não converja para a solução ideal. Já valores muito pequenos tornam o treinamento mais lento. No código foi utilizada uma taxa de aprendizado de 0,001, que é um valor comum por proporcionar estabilidade ao treinamento.

O clipvalue é um mecanismo de gradient clipping que limita o valor máximo dos gradientes durante a atualização dos pesos. Com clipvalue = 0,5, qualquer gradiente com magnitude superior a 0,5 é truncado para esse limite. Isso evita o problema de explosão dos gradientes (exploding gradients), tornando o treinamento mais estável e reduzindo oscilações excessivas na otimização.

In [ ]:
optimizer = tf.keras.optimizers.Adam(learning_rate = 0.001, clipvalue = 0.5)
neural_network.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['binary_accuracy'])

**12) Teste novamente a RNA. Aumentar a quantidade de camadas melhorou ou piorou os resultados? Explique o que aconteceu com a RNA e porque.**<br>
Ao aumentar a quantidade de camadas a acurácia caiu para 76% enquanto que a perda (loss) aumentou para aproximadamente 19.65, sou seja a rede teve resultados piores. Ao aumentar sua capacidade de aprendizado incluindo uma nova camada a rede também se tornou mais suscetível ao **overfitting** pois reduziu sua capacidade de generalização.

In [ ]:
neural_network.fit(X_train, y_train, batch_size=10, epochs=100)

Epoch 1/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 3s 26ms/step - binary_accuracy: 0.6338 - loss: 9924.2949 
Epoch 2/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.6033 - loss: 2230.2622
Epoch 3/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.5681 - loss: 372.1704
Epoch 4/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.5399 - loss: 223.7849
Epoch 5/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6033 - loss: 124.1847
Epoch 6/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6033 - loss: 66.5894
Epoch 7/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6432 - loss: 43.3404
Epoch 8/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6455 - loss: 30.7919
Epoch 9/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6925 - loss: 27.0050
Epoch 10/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5822 - loss: 54.4911
Epoch 11/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary

In [ ]:
predictions = neural_network.predict(X_test)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 52ms/step


In [ ]:
neural_network.evaluate(X_test, y_test)

5/5 ━━━━━━━━━━━━━━━━━━━━ 1s 56ms/step - binary_accuracy: 0.7622 - loss: 19.6598


[19.659839630126953, 0.7622377872467041]

## K-Fold Cross Validation

**13) Assista o vídeo https://youtu.be/RczbeFs_WbQ?si=f1Yu4LZaGDfZZpbZ para compreender melhor sobre como funciona o k-fold cross validation. Explique como esta rede foi configurada e como é possível chegar no resultado dela. O que é necessário fazer?**<br>
A rede é criada seguindo o mesmo padrão anterior, ou seja:
1. Uma camada de entrada com 30 neurônios
2. Duas camadas ocultas com 16 neurônios
3. Uma camada de saída que utiliza sigmoide como função de ativação

Além de utilizar o otimizador Adam, batch_size com valor 10 e 100 épocas de treinamento.

Em relação ao K-Fold é necessário chamar a função **cross_val_score** que recebe como paramêtro um "cv", este paramêtro é responsável por definir o total de folds, na implementação atual é definido cv=10. Portanto, em cada execução, 9 partes são utilizadas para treinamento e 1 parte é utilizada para validação. Esse processo é repetido 10 vezes, alternando a parte utilizada para teste, assim fazendo com que todas as amostras sejam avaliadas uma vez.

**14) Calcule também o Desvio Padrão dos resultados para avaliar o modelo. O que é possível concluir com esse resultado?**<br>

O desvio padrão calculado foi de 9,74%, isso significa que o modelo ainda não apresenta uma estabilidade ideal, o que justifica o uso da técnica de dropout que será utilizada a seguir.



In [ ]:
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import cross_val_score
from tensorflow.keras.models import Sequential
from tensorflow.keras import backend as k
from tensorflow.keras.layers import Dense, Dropout, InputLayer

In [ ]:
def create_network(dropout=False):
    k.clear_session()

    neural_network = Sequential()

    neural_network.add(InputLayer(shape=(30,)))

    neural_network.add(
        Dense(
            units=16,
            activation='relu',
            kernel_initializer='random_uniform'
        )
    )

    if dropout:
        neural_network.add(Dropout(0.2))

    neural_network.add(
        Dense(
            units=16,
            activation='relu',
            kernel_initializer='random_uniform'
        )
    )

    if dropout:
        neural_network.add(Dropout(0.2))

    neural_network.add(
        Dense(
            units=1,
            activation='sigmoid'
        )
    )

    optimizer = tf.keras.optimizers.Adam(
        learning_rate=0.001,
        clipvalue=0.5
    )

    neural_network.compile(
        optimizer=optimizer,
        loss='binary_crossentropy',
        metrics=['binary_accuracy']
    )

    return neural_network

In [ ]:
neural_network = KerasClassifier(model = create_network, epochs = 100, batch_size = 10)

In [ ]:
resultados = cross_val_score(estimator = neural_network, X = X, y = y, cv = 10, scoring = 'accuracy')

Epoch 1/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 6s 64ms/step - binary_accuracy: 0.6465 - loss: 0.6764
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - binary_accuracy: 0.7520 - loss: 0.4811
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.7891 - loss: 0.4994
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - binary_accuracy: 0.7949 - loss: 0.4759
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - binary_accuracy: 0.8398 - loss: 0.4620
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - binary_accuracy: 0.8125 - loss: 0.4745
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - binary_accuracy: 0.7949 - loss: 0.4743
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - binary_accuracy: 0.8379 - loss: 0.4417
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 1s 9ms/step - binary_accuracy: 0.8223 - loss: 0.5675
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - binary_accuracy: 0.8438 - loss: 0.4039
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8164 

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 60ms/step
Epoch 1/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 3s 19ms/step - binary_accuracy: 0.6152 - loss: 0.6868
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7129 - loss: 0.5878
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7715 - loss: 0.4928
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7461 - loss: 0.4901
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7578 - loss: 0.5516
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8262 - loss: 0.4703
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8320 - loss: 0.4206
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8438 - loss: 0.4576
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8457 - loss: 0.3987
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8574 - loss: 0.4695
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step
Epoch 1/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - binary_accuracy: 0.6191 - loss: 0.7470
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - binary_accuracy: 0.7422 - loss: 0.4925
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.7734 - loss: 0.4801
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8398 - loss: 0.4213
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8418 - loss: 0.4075
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8164 - loss: 0.5041
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8438 - loss: 0.4539
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8086 - loss: 0.4739
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8535 - loss: 0.3822
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8477 - loss: 0.4498
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 

In [ ]:
resultados

array([0.61403509, 0.8245614 , 0.78947368, 0.73684211, 0.77192982,
       0.92982456, 0.84210526, 0.9122807 , 0.68421053, 0.67857143])

In [ ]:
import numpy as np
standard_deviation = np.std(resultados)
standard_deviation

np.float64(0.09735884467927106)

## Overfitting e Dropout

**14) Como vimos, o modelo anterior sofreu overfitting. Para reduzir esse problema, aplicamos a técnica de regularização conhecida como dropout (artigo oficial: https://www.cs.toronto.edu/~rsalakhu/papers/srivastava14a.pdf e explicação em português: https://www.deeplearningbook.com.br/como-funciona-o-dropout/ ). Aplique o dropout de 20% na primeira e segunda camada oculta. O que acontece com os resultados? E o Desvio Padrão?**

In [ ]:
# layers.Dropout(rate=0.3)
model = create_network(dropout=True)
neural_network = KerasClassifier(model=model, epochs=100, batch_size=10)

NameError: name 'create_network' is not defined

In [ ]:
results = cross_val_score(estimator=neural_network, X=X, y=y, cv=10, scoring='accuracy')

Epoch 1/100


/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 32ms/step - binary_accuracy: 0.5703 - loss: 1.6407
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - binary_accuracy: 0.6113 - loss: 0.6802
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6719 - loss: 0.6042
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6426 - loss: 0.6209
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6426 - loss: 0.6464
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6719 - loss: 0.6132
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6719 - loss: 0.6138
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7031 - loss: 0.5440
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7227 - loss: 0.6571
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6953 - loss: 0.6210
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6816 - loss: 0.68

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 37ms/step - binary_accuracy: 0.5547 - loss: 1.1871
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6074 - loss: 0.7839
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6055 - loss: 0.6324
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6484 - loss: 0.5686
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6582 - loss: 0.6406
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6445 - loss: 0.5672
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6738 - loss: 0.5907
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7070 - loss: 0.6156
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6875 - loss: 0.7053
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7012 - loss: 0.5681
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7148 - loss: 0.56

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 28ms/step - binary_accuracy: 0.5527 - loss: 1.5072
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6152 - loss: 0.7407
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6348 - loss: 0.6338
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6484 - loss: 0.6382
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6367 - loss: 0.5743
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6602 - loss: 0.5890
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6777 - loss: 0.6216
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6738 - loss: 0.5893
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6816 - loss: 0.6375
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7109 - loss: 0.5508
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7109 - loss: 0.62

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 33ms/step - binary_accuracy: 0.5547 - loss: 1.2652
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5820 - loss: 0.7707
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6465 - loss: 0.6044
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6289 - loss: 0.5800
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6758 - loss: 0.5882
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6621 - loss: 0.5898
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6777 - loss: 0.6186
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7051 - loss: 0.5406
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7480 - loss: 0.5867
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7383 - loss: 0.5755
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7266 - loss: 0.59

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 32ms/step - binary_accuracy: 0.5352 - loss: 1.3336
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6582 - loss: 0.6428
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6523 - loss: 0.6081
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6816 - loss: 0.6195
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6523 - loss: 0.6822
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6719 - loss: 0.6032
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6992 - loss: 0.6426
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7051 - loss: 0.5924
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6953 - loss: 0.6495
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7324 - loss: 0.5056
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7207 - loss: 0.59

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 38ms/step - binary_accuracy: 0.5293 - loss: 1.3520
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6426 - loss: 0.6809
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6113 - loss: 0.6702
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6582 - loss: 0.5772
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6406 - loss: 0.6087
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6719 - loss: 0.5995
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6797 - loss: 0.5645
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6934 - loss: 0.5819
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6953 - loss: 0.6420
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6953 - loss: 0.6237
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7188 - loss: 0.54

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 38ms/step - binary_accuracy: 0.5137 - loss: 1.4947
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6543 - loss: 0.6538
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6562 - loss: 0.6473
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6504 - loss: 0.6283
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6641 - loss: 0.6836
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6660 - loss: 0.5675
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6973 - loss: 0.5759
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6836 - loss: 0.5855
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7070 - loss: 0.5799
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7246 - loss: 0.5848
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7441 - loss: 0.52

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 39ms/step - binary_accuracy: 0.5332 - loss: 1.3663
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - binary_accuracy: 0.5801 - loss: 0.7731
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6133 - loss: 0.6482
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6465 - loss: 0.6138
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6191 - loss: 0.6062
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6621 - loss: 0.5805
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6738 - loss: 0.5989
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6660 - loss: 0.5655
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6426 - loss: 0.6359
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6406 - loss: 0.5893
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6758 - loss: 0.62

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 26ms/step - binary_accuracy: 0.5410 - loss: 1.3022
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5801 - loss: 0.7356
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6211 - loss: 0.6126
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6875 - loss: 0.5820
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6328 - loss: 0.6571
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6758 - loss: 0.6124
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7168 - loss: 0.4904
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7168 - loss: 0.5960
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7441 - loss: 0.5528
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7422 - loss: 0.5508
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7402 - loss: 0.55

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 30ms/step - binary_accuracy: 0.5653 - loss: 1.4184
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5906 - loss: 0.7349
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6218 - loss: 0.6607
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6257 - loss: 0.6614
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6394 - loss: 0.7339
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6842 - loss: 0.5830
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6784 - loss: 0.7149
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6998 - loss: 0.6288
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.7173 - loss: 0.5954
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7524 - loss: 0.5916
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7446 - loss: 0.55

In [ ]:
results

array([0.70175439, 0.84210526, 0.92982456, 0.80701754, 0.92982456,
       0.9122807 , 0.92982456, 0.92982456, 0.85964912, 0.94642857])

In [ ]:
import numpy as np
standard_deviation = np.std(results)
standard_deviation

## Tuning dos Hiperparâmetros (técnicas de ajuste)

Hyperparameter Tuning (Ajuste de Hiperparâmetros) é o processo geral de encontrar os melhores valores para os hiperparâmetros de um modelo de machine learning.
Exemplos de hiperparâmetros:

- Taxa de aprendizado (learning rate)

- Número de árvores em um RandomForest

- Número de camadas ou neurônios em uma rede neural

- Tipo de kernel em um SVM

O tuning pode ser feito por diferentes técnicas, como:

- Grid Search

- Random Search

- Bayesian Optimization, etc.

**15) Descreva como a RNA foi configurada para fazer o processo de tuning.**
A rede é criada em partes de maneira igual, mesmas camadas e dropouts. Porém agora a função que cria a rede recebe os hiperparamêtros. Criamos **params** que serão testados utilizando Grid Search, optamos por utilizar a segunda versão do params que conta com menos opções pois a primeira acaba crashando a execução após algumas épocas de treinamento.<br>

Os melhores paramêtros encontrados foram: <br>
1. batch_size: 10
2. epochs: 50
3. model__activation: 'relu'
4. model__kernel_initializer: random_uniform
5. model__loss: binary_crossentropy
6. model__neurons: 16
7. model__optimizer: 'adam'

E que com eles foi possível obter uma acurácia de aproximadamente 89%.


In [ ]:
def create_network(optimizer, loss, kernel_initializer, activation, neurons):
  k.clear_session()
  rede_neural = Sequential([
      tf.keras.layers.InputLayer(shape=(30,)),
      tf.keras.layers.Dense(units=neurons, activation=activation, kernel_initializer=kernel_initializer),
      tf.keras.layers.Dropout(rate = 0.2),
      tf.keras.layers.Dense(units=neurons, activation=activation, kernel_initializer=kernel_initializer),
      tf.keras.layers.Dropout(rate = 0.2),
      tf.keras.layers.Dense(units=1, activation = 'sigmoid')])
  rede_neural.compile(optimizer = optimizer, loss = loss, metrics = ['binary_accuracy'])
  return rede_neural

In [ ]:
neural_network = KerasClassifier(model=create_network)

In [ ]:
params = {
    'batch_size': [10, 30],
    'epochs': [50, 100],
    'model__optimizer': ['adam', 'sgd'],
    'model__loss': ['binary_crossentropy', 'hinge'],
    'model__kernel_initializer': ['random_uniform', 'normal'],
    'model__activation': ['relu', 'tanh'],
    'model__neurons': [16, 8]
}

In [ ]:
#caso tiver problema com a configuração anterior, testar essa
params = {
    'batch_size': [10, 30],
    'epochs': [50],
    'model__optimizer': ['adam'],
    'model__loss': ['binary_crossentropy'],
    'model__kernel_initializer': ['random_uniform', 'normal'],
    'model__activation': ['relu'],
    'model__neurons': [16]
}

In [ ]:
params

{'batch_size': [10, 30],
 'epochs': [50],
 'model__optimizer': ['adam'],
 'model__loss': ['binary_crossentropy'],
 'model__kernel_initializer': ['random_uniform', 'normal'],
 'model__activation': ['relu'],
 'model__neurons': [16]}

In [ ]:
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(estimator = neural_network, param_grid = params,
                          scoring = 'accuracy', cv = 5)

In [ ]:
grid_search = grid_search.fit(X, y)

Epoch 1/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 4s 42ms/step - binary_accuracy: 0.5824 - loss: 1.0185
Epoch 2/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6308 - loss: 0.6474
Epoch 3/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7033 - loss: 0.5784
Epoch 4/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6703 - loss: 0.5682
Epoch 5/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7341 - loss: 0.4839
Epoch 6/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7407 - loss: 0.5177
Epoch 7/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7670 - loss: 0.4562
Epoch 8/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7780 - loss: 0.4682
Epoch 9/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7824 - loss: 0.4449
Epoch 10/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8286 - loss: 0.4106
Epoch 11/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8132 - loss: 0.3

In [ ]:
print(grid_search)

GridSearchCV(cv=5,
             estimator=KerasClassifier(model=<function create_network at 0x7d1fed61aac0>),
             param_grid={'batch_size': [10, 30], 'epochs': [50],
                         'model__activation': ['relu'],
                         'model__kernel_initializer': ['random_uniform',
                                                       'normal'],
                         'model__loss': ['binary_crossentropy'],
                         'model__neurons': [16], 'model__optimizer': ['adam']},
             scoring='accuracy')


In [ ]:
best_params = grid_search.best_params_
print(best_params)

{'batch_size': 10, 'epochs': 50, 'model__activation': 'relu', 'model__kernel_initializer': 'random_uniform', 'model__loss': 'binary_crossentropy', 'model__neurons': 16, 'model__optimizer': 'adam'}


In [ ]:
best_score = grid_search.best_score_
print(best_score)

0.8910883403198261


**16) É possível melhorar ainda mais a curácia da RNA? Como?**

## Para salvar o seu modelo de RNA

In [ ]:
best_model = grid_search.best_estimator_
best_model.model_.save('classificador_breast.keras')


In [ ]:
classificador_novo = tf.keras.models.load_model('/content/classificador_breast.keras')

In [ ]:
classificador_novo.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 16)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 16)             │           272 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 16)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,357 (9.21 KB)

 Trainable params: 785 (3.07 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 1,572 (6.14 KB)

# PARTE 2: Redes Neurais e CNN - Identificação de Objetos

Neste trabalho iremos fazer a identificação de um conjunto de dados de imagens de artigos de Zalando chamado Fashion-MNIST, que consiste em um conjunto de treinamento de 60.000 exemplos e um conjunto de teste de 10.000 exemplos. Cada exemplo é uma imagem em tons de cinza 28x28, associada a um rótulo de 10 classes.

Fonte: https://www.tensorflow.org/datasets/catalog/fashion_mnist?hl=pt-br

## **ATIVIDADE:**

**Descrição:**

Nesta atividade, você receberá um código de uma Rede Neural Artificial e uma Rede Neural Convolucional (CNN) implementada. Sua tarefa é analisar o código fornecido e fazer uma interpretação detalhada de cada linha, identificando e explicando as principais etapas, camadas e operações realizadas pela RNA e pela CNN.

**Instruções:**

1. Preencha o código com comentários em cada linha, explicando sua função e propósito.
2. Analise o código cuidadosamente para entender como a CNN é configurada.
3. Identifique e explique as camadas de convolução, camadas de pooling, camadas densas, funções de ativação, tamanhos dos filtros, número de neurônios, etc.
4. Identifique quais técnicas de regularização ou otimização formam utilizadas no código e explique como elas contribuem para o desempenho e a generalização do modelo.
5. Comente sobre a função de ativação utilizada na camada de saída.
6. Descreva os hiperparâmetros e ajustes do modelo, como a taxa de aprendizado, número de épocas de treinamento e o tratamento dos conjuntos de dados de treinamento e teste.
7. Faça uma avaliação descritiva dos resultados das Redes Neurais implementadas.



**Entregável:**

Prepare uma análise detalhada do código fornecido, destacando as principais características da CNN e como ela é configurada para a tarefa em questão. Se necessário, inclua observações sobre o potencial de melhoria ou otimização do modelo.

Esta atividade visa consolidar seu conhecimento sobre CNNs, ajudando a compreender como uma Rede Neural Convolucional é implementada e configurada para tarefas específicas.

In [ ]:
import tensorflow as tf
from tensorflow import keras

In [ ]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
%matplotlib inline

## Importando base de dados

In [ ]:
fashion_mnist = keras.datasets.fashion_mnist


In [ ]:
(X_train_full, y_train_full), (X_test, y_test) = fashion_mnist.load_data()

In [ ]:
print(X_train_full.shape)
print(y_train_full.shape)
print(X_test.shape)
print(y_test.shape)

In [ ]:
# Transformar a escala para que os valores fiquem entre 0 e 1
X_valid, X_train = X_train_full[:5000] / 255.0, X_train_full[5000:] / 255.0
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

In [ ]:
nomes_classes = ["camisa/top", "calca", "casaco", "vestido", "jaqueta",
               "sandalia", "camiseta", "tenis", "bolsa", "bota"]

In [ ]:
nomes_classes[y_train[5]]

In [ ]:
X_train[5]

In [ ]:
exemplo_n = 5
print(nomes_classes[y_train[exemplo_n]])
plt.imshow(X_train[exemplo_n], interpolation='nearest')
plt.show()

##  Rede Neural Artificial com Keras

In [ ]:
nn = keras.models.Sequential()
nn.add(keras.layers.Flatten(input_shape=[28, 28]))
nn.add(keras.layers.Dense(300, activation="relu"))
nn.add(keras.layers.Dropout(rate=0.1))
nn.add(keras.layers.Dense(100, activation="relu"))
nn.add(keras.layers.Dropout(rate=0.1))
nn.add(keras.layers.Dense(10, activation="softmax"))
nn.summary()

### Compilando e treinando o modelo

In [ ]:
%time
nn.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])
history_nn = nn.fit(X_train, y_train, epochs=5, validation_data=(X_valid, y_valid))

### Visualizar a performance

In [ ]:
pd.DataFrame(history_nn.history).plot(figsize=(12, 8))
plt.grid(True)
plt.gca().set_ylim(0, 1) # set the vertical range to [0-1]
plt.show()

### Avaliar o modelo

In [ ]:
# prints the loss and the accuracy
nn.evaluate(X_test, y_test,verbose=1)

###Realizar uma Previsão

In [ ]:
X_novo = X_test[:4]
y_proba = nn.predict(X_novo)
y_proba

In [ ]:
y_pred = np.argmax(nn.predict(X_novo), axis=-1)
print(y_pred)
print(np.array(nomes_classes)[y_pred])

### Verificar os resultados

In [ ]:
y_novo = y_test[:4]
print(y_novo)
np.array(nomes_classes)[y_pred]

In [ ]:
# example_n = 201
print(nomes_classes[y_pred[2]])
plt.imshow(X_novo[2], interpolation='nearest')
plt.show()

## Treinando CNN com Keras

In [ ]:
cnn = keras.models.Sequential([
keras.layers.InputLayer(input_shape = (28,28,1)),
keras.layers.Conv2D(64,5, activation='relu', padding='same', kernel_initializer='glorot_uniform'),
keras.layers.MaxPool2D(2),
keras.layers.Conv2D(128,3, activation='relu', padding='same'),
keras.layers.Conv2D(128,3, activation='relu', padding='same'),
keras.layers.MaxPool2D(2),
keras.layers.Conv2D(258,3, activation='relu', padding='same'),
keras.layers.Conv2D(258,3, activation='relu', padding='same'),
keras.layers.MaxPool2D(2),
keras.layers.Flatten(),
keras.layers.Dense(128, activation="relu"),
keras.layers.Dropout(rate=0.5),
keras.layers.Dense(64, activation="relu"),
keras.layers.Dropout(rate=0.5),
keras.layers.Dense(10, activation="softmax")])


In [ ]:
cnn.summary()

In [ ]:
print(X_train.shape)
X_train_new = X_train.reshape(X_train.shape[0], X_train.shape[1], X_train.shape[2],1)
X_valid_new = X_valid.reshape(X_valid.shape[0], X_valid.shape[1], X_valid.shape[2],1)
X_test_new = X_test.reshape(X_test.shape[0], X_test.shape[1], X_test.shape[2],1)
print(X_train_new.shape)

In [ ]:
cnn.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])
history_cnn = cnn.fit(X_train_new, y_train, epochs=20, validation_data=(X_valid_new, y_valid))

### Avaliar o modelo

In [ ]:
cnn.evaluate(X_test_new, y_test,verbose=1)

In [ ]:
pd.DataFrame(history_cnn.history).plot(figsize=(12, 8))
plt.grid(True)
plt.gca().set_ylim(0, 1) # set the vertical range to [0-1]
plt.show()